sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateFormatter.scala

```diff
@@ -17,24 +17,30 @@
 
 package org.apache.spark.sql.catalyst.util
 
-import java.time.LocalDate
+import java.time.{LocalDate, ZoneId}
 import java.util.Locale
 
+import DateTimeUtils._
+
 sealed trait DateFormatter extends Serializable {
   def parse(s: String): Int // returns days since epoch
   def format(days: Int): String
 }
 
 class Iso8601DateFormatter(
     pattern: String,
+    zoneId: ZoneId,
     locale: Locale) extends DateFormatter with DateTimeFormatterHelper {
 
   @transient
   private lazy val formatter = getOrCreateFormatter(pattern, locale)
 
   override def parse(s: String): Int = {
-    val localDate = LocalDate.parse(s, formatter)
-    DateTimeUtils.localDateToDays(localDate)
+    val specialDate = convertSpecialDate(s.trim, zoneId)
```

Comment:
We always trim input inside of `stringToDate`. Maybe here we should trim the input in all cases? /cc @HyukjinKwon @srowen

Reply:
Since casts from strings to date/tmestamp already does so, the trim looks reasonable to me.

Reply:
One thing I am a bit worried though, if users set the date `pattern` including the leading or trailing white spaces intentionally (because they know there's spaces), it will now returns invalid results.

```scala
DateFormatter(" yyyy MMM").parse("2018 Dec")
```
```
Text '2018 Dec' could not be parsed at index 0
java.time.format.DateTimeParseException: Text '2018 Dec' could not be parsed at index 0
	at java.base/java.time.format.DateTimeFormatter.parseResolved0(Unknown Source)
	at java.base/java.time.format.DateTimeFormatter.parse(Unknown Source)
	at java.base/java.time.LocalDate.parse(Unknown Source)
	at org.apache.spark.sql.catalyst.util.Iso8601DateFormatter.parse(DateFormatter.scala:36)
	at org.apache.spark.sql.util.DateFormatterSuite.$anonfun$new$15(DateFormatterSuite.scala:94)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.java:23)
```

While trimming makes sense in general, there's difference comparing to `stringToDate` since here we allow to set the `pattern` (and with this Java 8 APIs, now Spark can support exact and better match with the pattern).

Is this change necessary or does this change make it easier to fix the issue?

Reply:
@HyukjinKwon I just asked because I am doing trimming here for checking special values. So, the overhead already exists. Maybe, it could be useful in another branch.